eth/fetcher/tx_fetcher.go

```diff
@@ -0,0 +1,291 @@
+	op := &txInject{
+		origin: peer,
+		txs:    txs,
+	}
+	select {
+	case f.inject <- op:
+		return nil
+	case <-f.quit:
+		return errTerminated
+	}
+}
+
+// Start boots up the announcement based synchroniser, accepting and processing
+// hash notifications and block fetches until termination requested.
+func (f *TxFetcher) Start() {
+	go f.loop()
+}
+
+// Stop terminates the announcement based synchroniser, canceling all pending
+// operations.
+func (f *TxFetcher) Stop() {
+	close(f.quit)
+}
+
+func (f *TxFetcher) loop() {
+	fetchTimer := time.NewTimer(0)
+
+	for {
+		// Clean up any expired transaction fetches.
+		// There are many cases can lead to it:
+		// * We send the request to busy peer which can reply immediately
+		// * We send the request to malicious peer which doesn't reply deliberately
+		// * We send the request to normal peer for a batch of transaction, but some
+		//   transactions have been included into blocks. According to EIP these txs
+		//   won't be included.
+		// But it's fine to delete the fetching record and reschedule fetching iff we
+		// receive the annoucement again.
+		for hash, announce := range f.fetching {
+			if time.Since(announce.time) > txFetchTimeout {
+				delete(f.fetching, hash)
+				txFetchTimeoutMeter.Mark(1)
+			}
+		}
+		select {
+		case anno := <-f.notify:
+			txAnnounceInMeter.Mark(1)
+
+			// Drop the new announce if there are too many accumulated.
+			count := f.announces[anno.origin] + 1
+			if count > txAnnounceLimit {
+				txAnnounceDOSMeter.Mark(1)
+				break
+			}
+			// All is well, schedule the annouce if transaction is not yet downloading
+			if _, ok := f.fetching[anno.hash]; ok {
+				break
+			}
+			f.announces[anno.origin] = count
+			f.announced[anno.hash] = append(f.announced[anno.hash], anno)
+
+			if len(f.announced) == 1 {
+				f.reschedule(fetchTimer)
+			}
+			if f.announceHook != nil {
+				f.announceHook(anno.hash)
+			}
+		case <-fetchTimer.C:
+			// At least one tx's timer ran out, check for needing retrieval
+			request := make(map[string][]common.Hash)
+
+			for hash, announces := range f.announced {
+				if time.Since(announces[0].time) > arriveTimeout-gatherSlack {
+					// Pick a random peer to retrieve from, reset all others
+					announce := announces[rand.Intn(len(announces))]
+					f.forgetHash(hash)
+
+					// Skip fetching if we already receive the transaction.
+					if f.hasTx(hash) {
+						txAnnounceSkipMeter.Mark(1)
+						continue
+					}
+					// If the transaction still didn't arrive, queue for fetching
+					request[announce.origin] = append(request[announce.origin], hash)
+					f.fetching[hash] = announce
+				}
+			}
+			// Send out all block header requests
+			for peer, hashes := range request {
+				log.Trace("Fetching scheduled transactions", "peer", peer, "txs", hashes)
+
+				// Create a closure of the fetch and schedule in on a new thread
+				fetchTxs, hashes := f.fetching[hashes[0]].fetchTxs, hashes
+
+				go func() {
```

Comment:
I'm unsure if it's wise to spin up a new goroutine here. I could imagine someone announcing stuff and not reading the requests afterwards, causing these goroutines to pile up. Perhaps we should add an `AsyncGetPooledTransactions` method too which would use the same `peer.broadcast` to ensure we don't go crazy.

If we were to do that, we could dump arbitrarily many hashes there and have the `broadcast` pick `MaxTransactionFetch` number per request to send over.

Another question is whether we'd like to enforce one single retrieval to be in flight at the same time or allow arbitrary (currently we fire as many as it takes to get all the hashes across).